src/decorators.ts

```diff
@@ -0,0 +1,157 @@
+ * UpdatingElement as exported by the lit-element module. That means it *must*
+ * be used in conjunction with the @property decorator, or equivalent
+ * JavaScript.
+ *
+ * Supported configurations are:
+ *
+ *  - `intrinsics`: An Intrinsics struct that describes how to interpret a
+ * serialized style attribute. For more detail on intrinsics see
+ * ./styles/evaluators.ts
+ *  - `updateHandler`: A string or Symbol that is the key of a method to be
+ * invoked with the result of parsing and evaluating a serialized style string.
+ *  - `observeEffects`: Optional, if set to true then styles that use env() will
+ * cause their update handlers to be invoked every time the corresponding
+ * environment variable changes (even if the style attribute itself remains
+ * static).
+ */
+export const style =
+    <T extends Intrinsics<Array<Unit>>, U extends UpdatingElement>(
+        config: StyleDecoratorConfig<T, U>) => {
+      const observeEffects: boolean = config.observeEffects || false;
+      const getIntrinsics = config.intrinsics instanceof Function ?
+          config.intrinsics :
+          (() => config.intrinsics) as IntrinsicsFactory<T, U>;
+
+      return <U extends typeof UpdatingElement['prototype']>(
+                 proto: U, propertyName: string) => {
+        const originalUpdated = (proto as any).updated;
+        const originalConnectedCallback = proto.connectedCallback;
+        const originalDisconnectedCallback = proto.disconnectedCallback;
+
+        const $styleEffector = Symbol(`${propertyName}StyleEffector`);
+        const $styleEvaluator = Symbol(`${propertyName}StyleEvaluator`);
+        const $updateEvaluator = Symbol(`${propertyName}UpdateEvaluator`);
+        const $evaluateAndSync = Symbol(`${propertyName}EvaluateAndSync`);
+
+        Object.defineProperties(proto, {
+          [$styleEffector]:
+              {value: null as StyleEffector | null, writable: true},
+          [$styleEvaluator]:
+              {value: null as StyleEvaluator<T>| null, writable: true},
+
+          [$updateEvaluator]: {
+            value: function() {
+              const ast = parseExpressions(
+                  this[propertyName as keyof UpdatingElement] as string);
+              this[$styleEvaluator] =
+                  new StyleEvaluator(ast, getIntrinsics(this));
+
+              if (this[$styleEffector] == null && observeEffects) {
+                this[$styleEffector] =
+                    new StyleEffector(() => this[$evaluateAndSync]());
+              }
+
+              if (this[$styleEffector] != null) {
+                this[$styleEffector].observeEffectsFor(ast);
+              }
+            }
+          },
+
+          [$evaluateAndSync]: {
+            value: function() {
+              if (this[$styleEvaluator] == null) {
+                return;
+              }
+
+              const result = this[$styleEvaluator].evaluate();
+
+              (this as unknown as
```

Comment:
don't ask me hard questions I just work here

Seriously though, this is the only way I could figure out to do this kind of dynamic dispatch "correctly." I collaborated with our local TypeScript gurus on this one and the recommendation was that there may be a more sound alternative but it will be extremely difficult to write correctly.